src/org/exist/xslt/EXistURIResolver.java

```diff
@@ -0,0 +1,201 @@
+
+    final String[] pathComponents = path.substring(1).split("/");
+
+    final int numPathComponents = Array.getLength(pathComponents);
+    final String[] simplifiedComponents = new String[numPathComponents];
+    int numSimplifiedComponents = 0;
+
+    for (final String s : pathComponents) {
+      // Remove empty elements ("/")
+      if (s.isEmpty()) {
+        continue;
+      }
+      // Remove identity elements (".")
+      if (".".equals(s)) {
+        continue;
+      }
+      // Remove parent elements ("..") unless at the root
+      if ("..".equals(s)) {
+        if (numSimplifiedComponents > 0) {
+          numSimplifiedComponents--;
+        }
+        continue;
+      }
+      simplifiedComponents[numSimplifiedComponents++] = s;
+    }
+
+    if (numSimplifiedComponents == 0) {
+      return "/";
+    }
+
+    final StringBuilder sb = new StringBuilder(path.length());
+    for (int x = 0; x < numSimplifiedComponents; x++) {
+      sb.append("/").append(simplifiedComponents[x]);
+    }
+
+    if (path.endsWith("/")) {
+      sb.append("/");
+    }
+
+    return sb.toString();
+  }
+
+  @Override
+  public Source resolve(final String href, String base) throws TransformerException {
+    String path;
+
+    if (href.isEmpty()) {
+      path = base;
+    } else {
+      URI hrefURI = null;
+      try {
+        hrefURI = new URI(href);
+      } catch (final URISyntaxException e) {
+      }
+      if (hrefURI != null && hrefURI.isAbsolute()) {
+        path = href;
+      } else {
+        if (href.startsWith("/")) {
+          path = href;
+        } else if (href.startsWith(XmldbURI.EMBEDDED_SERVER_URI_PREFIX)) {
+          path = href.substring(XmldbURI.EMBEDDED_SERVER_URI_PREFIX.length());
+        } else if (base == null || base.length() == 0) {
+          path = basePath + "/" + href;
+        } else {
+          // Maybe base never contains this prefix?  Check to be sure.
+          if (base.startsWith(XmldbURI.EMBEDDED_SERVER_URI_PREFIX)) {
+            base = base.substring(XmldbURI.EMBEDDED_SERVER_URI_PREFIX.length());
+          }
+          path = base.substring(0, base.lastIndexOf("/") + 1) + href;
+        }
+      }
+    }
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Resolving path " + href + " with base " + base + " to " + path);
+      // + " (URI = " + uri.toASCIIString() + ")");
+    }
+
+    if (path.startsWith("/")) {
+      path = normalizePath(path);
+      return databaseSource(path);
+    } else {
+      return urlSource(path);
+    }
+  }
+
+  private Source urlSource(final String path) throws TransformerException {
+    try {
+      final URL url = new URL(path);
+      return new StreamSource(url.openStream());
+    } catch (final IOException e) {
+      throw new TransformerException(e.getMessage(), e);
+    }
+  }
+
+  private Source databaseSource(final String path) throws TransformerException {
+    final XmldbURI uri = XmldbURI.create(path);
+
+    final DBBroker broker = db.getActiveBroker();
```

Comment:
why not? No needs to allocate broker because it must be already allocated for processing.